Sources/NIOHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,267 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIOHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Swift Server Working Group and the SwiftNIOHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIOHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOTLS
+import NIOHTTP1
+import NIOHTTP2
+import NIOConcurrencyHelpers
+import Foundation
+
+class ConnectionPool {
+    
+    init(group: EventLoopGroup, configuration: HTTPClient.Configuration) {
+        self.eventLoopGroup = group
+        self.configuration = configuration
+    }
+    
+    var eventLoopGroup: EventLoopGroup
+    var configuration: HTTPClient.Configuration
+    private var connectionProviders: [Key:ConnectionProvider] = [:]
+    private var lock = Lock()
+    
+    private subscript(key: Key) -> ConnectionProvider? {
```

Comment:
I felt like it was better to make a function that would always warp gets/sets in lock since it's the main data structure that needs to be protected, the goal here is to reduce risks of threading issues and making less boilerplate.